pkg/operator/mon/mon.go

```diff
@@ -36,187 +39,399 @@ const (
 	monSecretName     = "mon-secret"
 	adminSecretName   = "admin-secret"
 	clusterSecretName = "cluster-name"
+	monConfigMapName  = "mon-config"
+	monEndpointKey    = "endpoints"
+	maxMonIDKey       = "maxMonId"
 )
 
 type Cluster struct {
 	Namespace    string
 	Keyring      string
-	ClusterName  string
 	Version      string
 	MasterHost   string
 	Size         int
 	Paused       bool
 	AntiAffinity bool
 	Port         int32
+	clientset    kubernetes.Interface
 	factory      client.ConnectionFactory
 	retryDelay   int
+	maxRetries   int
+	clusterInfo  *mon.ClusterInfo
+	maxMonID     int
+	configDir    string
+	waitForStart bool
 }
 
 type MonConfig struct {
 	Name string
 	Port int32
 }
 
-func New(namespace string, factory client.ConnectionFactory, version string) *Cluster {
+func New(clientset kubernetes.Interface, factory client.ConnectionFactory, namespace string, version string) *Cluster {
 	return &Cluster{
+		clientset:    clientset,
+		factory:      factory,
 		Namespace:    namespace,
 		Version:      version,
 		Size:         3,
-		factory:      factory,
 		AntiAffinity: true,
 		retryDelay:   6,
+		maxRetries:   15,
+		maxMonID:     -1,
+		configDir:    k8sutil.DataDir,
+		waitForStart: true,
 	}
 }
 
-func (c *Cluster) Start(clientset kubernetes.Interface) (*mon.ClusterInfo, error) {
+func (c *Cluster) Start() (*mon.ClusterInfo, error) {
 	logger.Infof("start running mons")
 
-	clusterInfo, err := c.initClusterInfo(clientset)
+	err := c.initClusterInfo()
 	if err != nil {
 		return nil, fmt.Errorf("failed to initialize ceph cluster info. %+v", err)
 	}
-	mons := []*MonConfig{}
-	for i := 0; i < c.Size; i++ {
-		mons = append(mons, &MonConfig{Name: fmt.Sprintf("mon%d", i), Port: int32(mon.Port)})
+
+	if len(c.clusterInfo.Monitors) == 0 {
+		// Start the initial monitors at startup
+		mons := c.getExpectedMonConfig()
+		err = c.startPods(nil, mons)
+		if err != nil {
+			return nil, fmt.Errorf("failed to start mon pods. %+v", err)
+		}
+	} else {
+		// Check the health of a previously started cluster
+		err = c.CheckHealth()
+		if err != nil {
+			logger.Warningf("failed to check mon health %+v. %+v", c.clusterInfo.Monitors, err)
+		}
 	}
 
-	err = c.startPods(clientset, clusterInfo, mons)
+	return c.clusterInfo, nil
+}
+
+func (c *Cluster) CheckHealth() error {
+	// update the config map if the pod ips changed
+	// must retry since during startup of pods they might take some time to initialize
+	k8sutil.Retry(time.Duration(c.retryDelay)*time.Second, c.maxRetries, func() (bool, error) {
+		err := c.updateConfigMapIfPodIPsChanged()
```

Comment:
as we talked about, let's add a comment here about how this won't result in a healthy cluster still after a reboot of all the mon hosts.